### ISSUE TITLE
feat: Keep track of the state (height, round, round step) of all validators

### ISSUE THREAD

**Issue by cason** [circlefin/malachite#316]
It would be useful for logging and debugging purposes.

It could also be helpful as an input for the node synchronization protocols (#260) to be designed.

In summary, the node should keep track of the latest state of other nodes in the network, which is inferred from the messages we receive from them. The state, a priori, is a triple `height, round, round_step`, i.e., three counters, where `round_step` is one of `proposal`, `prevote`, and `precommit`.

Whenever we receive a vote for a node `p` we update `state[p]` if the received message is "greater" than the current state. So, if the height of the message is higher than the registered height, or heights are the same but the round is higher, or if height and round are the same but the round step has increased (derived from the message type, `Proposal`, `Prevote`, or `Precommit`).

**Comment by cason** [comment]
Do we still need or are considering this approach? This could let to important optimisations for the synchronization protocols. Namely, we can ask the right nodes/validators for information they should have (e.g., if a node is in height 10, it must have the decisions for all heights up to 9).

Even if this is not the case, this summary of other nodes' state is useful for debugging.

**Comment by romac** [comment]
Yes would still be good to keep track of this somewhere. Just not sure how to infer the round step another node is at without explicitly publishing to the network, which I'd like to avoid adding at this stage to not introduce protocol changes. Do you see an alternative?

**Comment by ancazamfir** [comment]
> Do we still need or are considering this approach? This could let to important optimisations for the synchronization protocols. Namely, we can ask the right nodes/validators for information they should have (e.g., if a node is in height 10, it must have the decisions for all heights up to 9).
> 
> Even if this is not the case, this summary of other nodes' state is useful for debugging.

We don't keep certificates for past heights in consensus. They are kept at application side. We currently use those in value sync, any other usage you have in mind? 
Consensus may verify incoming consensus messages for current and higher heights for which it knows the validator set and update peer state based on that. But the horizon is limited to max 2 epochs.

**Comment by cason** [comment]
> Just not sure how to infer the round step another node is at without explicitly publishing to the network

From a validator we can infer it from the messages we receive from (signed by) them.

I think this approach loses a lit of meaning if we don't have a mapping between nodes and validators.
